### PR TITLE
Breaking changes for 2.0

### PR DESCRIPTION
**Pull request by Zandor300**
Fixes https://github.com/dennisweissmann/DeviceKit/issues/149
Closes https://github.com/dennisweissmann/DeviceKit/issues/130
Fixes https://github.com/dennisweissmann/DeviceKit/issues/167
Fixes https://github.com/dennisweissmann/DeviceKit/issues/165
Closes https://github.com/devicekit/DeviceKit/issues/174 (Both Swift 5 and the `@unknown` in switch)
Closes https://github.com/devicekit/DeviceKit/issues/159

---

Fixes https://github.com/dennisweissmann/DeviceKit/pull/164 not being defined in the gyb file.
Fixes https://github.com/dennisweissmann/DeviceKit/pull/166 not being defined in the gyb file.

---

# Changelog
## Breaking changes
- The original `Device()` constructor has been made private in favour of using `Device.current` to match `UIDevice.current`.
- The enum values for the iPhone Xs, iPhone Xs Max and iPhone Xʀ have been renamed to be `.iPhoneXS`, `.iPhoneXSMax` and `.iPhoneXR` to match proper formatting.
- `.description` for the iPhone Xs, iPhone Xs Max and iPhone Xʀ have been changed to contain small caps formatting for the s and the ʀ part.
- `.description` for the iPad 5 and iPad 6 have been changed to the proper names; iPad (5th generation) and iPad (6th generation).
- `.name`, `.systemName`, `.systemVersion`, `.model`, `.localizedModel`, `.batteryState` and `.batteryLevel` will now all return nil when you try to get its value when the device you are getting it from isn't the current one. (eg. `Device.iPad6.name` while running on iPad 5)

## New features
- Updated to Swift 5!
- New `.allDevicesWithRoundedDisplayCorners` and `.hasRoundedDisplayCorners` values to check if a device has rounded display corners. (eg. iPhone Xs and iPad Pro (3rd generation))
- new `.allDevicesWithSensorHousing` and `.hasSensorHousing` values to check if a device has a screen cutout for the sensor housing. (eg. iPhone Xs)

## Bugfixes
- `.isPad` and `.isPhone` are now giving correct outputs again.

### CODE REVIEW

**Comment by dennisweissmann** [comment]
Wow thank you for that PR :)

Thanks for the fix of #164, that looks awesome, totally missed that yesterday, that’s definitely the way to go!

This is also a todo list for me since I’m mobile only this month and I don’t know off the top of my head:

- are `Device.instance` and `Device.current` the same thing? If so lets either remove `.current` since `init()` already does the right thing (right?!) or go 100% your approach and replace `.instance` with `.current`
- thanks for fixing the cases, that’s great! (Should we use the “small capital letters” R and S if at all possible?
- should those early-return-if’s be guards?

Thank you for working on this!

**Comment by Zandor300** [comment]
1. I think `.current` indicates better that it is the current device the code is running on. `.instance` is currently private anyway.
2. I think there might be problems with some peoples applications if we move to using Xʀ and Xs but I might be wrong. I think it is a cool feature but maybe having a separate String variable for getting the small capital name may prevent problems?
3. That will definitely be cleaner looking 👍

**Comment by Zandor300** [comment]
Just changed it to using small caps in https://github.com/dennisweissmann/DeviceKit/pull/168/commits/32979363568c32ccece997eb157e1cc497515465.

**Comment by Zandor300** [comment]
Also slapped in those guards: https://github.com/dennisweissmann/DeviceKit/pull/168/commits/d8b5102189b074fd92a9d71fd4402f57bef33858

**Comment by Zandor300** [comment]
Just changed everything over to solely `.current`, you may look at it if you like it and if you don't just change my commit: https://github.com/dennisweissmann/DeviceKit/pull/168/commits/05cac804ca341f442dea9adbc9ca6b6f8afcabab

**Comment by Zandor300** [comment]
Just noticed; Small caps **s** is just the same as a lowercase **s**.

**Comment by dennisweissmann** [comment]
You are awesome! 😎 that is super cool Zander! I really like it :)

**Comment by Zandor300** [comment]
@dennisweissmann fixed my other open pull request also being defined through the generated swift file directly instead of using the gyb file, and fixed the documentation inside the readme and throughout the code, so that it is using `Device.current` instead of `Device()`

**Comment by dennisweissmann** [comment]
I just made you a collaborator Zandor :) I’ll hand over the project to my colleague later this week (I’m not sure if we’ll move it to her account or create an org, I just wanted to thank you for your continuous work and passion for the project, being a collaborator should give you (hopefully) merge rights and more! :) 🙏

**Comment by Zandor300** [comment]
@dennisweissmann Thanks for that, I now have push access. 👍 I will use it with caution! I +1 creating an organisation.

**Comment by Zandor300** [comment]
@dennisweissmann I've added a changelog which can be used for when these changes will be released in the future.

**Comment by dennisweissmann** [comment]
That’s awesome! Always wanted to have one!

**Comment by Zandor300** [comment]
@dennisweissmann @denisenepraunig Offtopic: Can we like set up a Discord server or WhatsApp for more direct communication so I can reach you?

I think it is smart to release a 1.12.1 or 1.13 version for the new iPads before we update to Swift 5. That way, people that don't want to update yet but do want the new iPads can still have those changes.

**Comment by Zandor300** [comment]
Updated to Swift 5! 🎉

**Comment by Zandor300** [comment]
Ready for merging in my opinion. Still think we should release a 1.13.0 version of the `master` branch as it is currently to release the new iPads for Swift 4.2. Then the `master` branch should be merged into the `swift-4.2` branch before this pull request is merged into `master`.

File: Source/Device.generated.swift [hunk 1]
```diff
@@ -617,9 +615,10 @@ public enum Device {
       return Device.realDevice(from: self)
     }
 
-    public var isZoomed: Bool {
+    public var isZoomed: Bool? {
+      guard isCurrent else { return nil }
       // TODO: Longterm we need a better solution for this!
```
[the remark this answers is not in the document]
Comment (Zandor300) [reply Source/Device.generated.swift]: Duplicate

File: Source/Device.generated.swift [hunk 2]
```diff
@@ -617,9 +615,10 @@ public enum Device {
       return Device.realDevice(from: self)
     }
 
-    public var isZoomed: Bool {
+    public var isZoomed: Bool? {
```
[the remark this answers is not in the document]
Comment (Zandor300) [reply Source/Device.generated.swift]: Duplicate

**Comment by dennisweissmann** [comment]
We can set something like this up, sure, but since I won’t have the time to contribute or manage very much going forward, I’d like to leave that up to @denisenepraunig

**Comment by denisenepraunig** [comment]
@Zandor300 I am fine with WhatsApp or Telegram for direct communication - for questions I have to Dennis I will contact him directly. You can drop me a mail with your number to denise dot nepraunig at gmail dot com.

**Comment by robbiet480** [comment]
Another warning that needs fixing:

```
/Users/robbiet480/Repos/HomeAssistant/home-assistant-iOS/Pods/DeviceKit/Source/Device.generated.swift:994:9: Switch covers known cases, but 'WKInterfaceDeviceBatteryState' may have additional unknown values, possibly added in future versions
```

**Comment by Zandor300** [comment]
@robbiet480 Thanks for the catch! Fixed it in https://github.com/devicekit/DeviceKit/pull/168/commits/11353eb819a7a5738f5b10f56fbb911949bc96d6.

**Comment by robbiet480** [comment]
@denisenepraunig LGTM!

**Pull request merged**